Testing.Rmd

```diff
@@ -275,140 +382,640 @@ All expectations have a similar structure:
 
 * They start with `expect_`.
 
-* They have two arguments: the first is the actual result, the second is what
-  you expect.
+* They have two main arguments:
+  the first is the actual result, the second is what you expect.
   
 * If the actual and expected results don't agree, testthat throws an error.
 
+* Some expectations have additional arguments that control the finer points of
+  comparing an actual and expected result.
+
 While you'll normally put expectations inside tests inside files, you can also run them directly.
 This makes it easy to explore expectations interactively.
-There are almost 20 expectations in the testthat package.
-The most important are discussed below.
+There are more than 40 expectations in the testthat package, which can be explored in testthat's [reference index](https://testthat.r-lib.org/reference/index.html).
+Here we highlight some of the most important expectations.
 
-*This passage is still being revised for the 2nd edition the R Packages book.
-In the meantime, please note that:
-in the third edition of testthat, `expect_equal()` and `expect_identical()` are both powered by the [waldo package]().
-In the past / in the second edition of testthat, `expect_equal()` wraps `all.equal()` and `expect_identical()` wraps `identical()`.*
-
-*   There are two basic ways to test for equality: `expect_equal()`, 
-    and `expect_identical()`. `expect_equal()` is the most commonly used: it 
-    checks for equality within a numerical tolerance:
+* There are two basic ways to test for equality: `expect_equal()`, and
+  `expect_identical()`.
+  `expect_equal()` is the most commonly used and should be your default:
+  it checks for equality within a numerical tolerance:
 
-    ```{r, error = TRUE, eval = FALSE}
+    ```{r, error = TRUE}
     expect_equal(10, 10)
+    expect_equal(10, 10L)
     expect_equal(10, 10 + 1e-7)
     expect_equal(10, 11)
     ```
   
-    If you want to test for exact equivalence, or need to compare a more
-    exotic object like an environment, use `expect_identical()`.
+  If you want to test for exact equivalence or need to compare more exotic
+  objects like environments, use `expect_identical()`.
 
-    ```{r, error = TRUE, eval = FALSE}
+    ```{r, error = TRUE}
     expect_equal(10, 10 + 1e-7)
     expect_identical(10, 10 + 1e-7)
+    expect_equal(2, 2L)
+    expect_identical(2, 2L)
     ```
 
-*   `expect_match()` matches a character vector against a regular expression. 
-    The optional `all` argument controls whether all elements or just one 
-    element needs to match.
-    This is powered by `grepl()` (additional arguments like
-    `ignore.case = FALSE` or `fixed = TRUE` are passed on down).
-
-    ```{r, error = TRUE, eval = FALSE}
+* `expect_match()` matches a character vector against a regular expression. 
```

Comment:
I think this should come later or maybe not at all?

Maybe there should be (or already is) a section that lists a bunch of shortcuts, e,g:

* `expect_length(x, n)` is a short cut for `expect_equal(length(x), n)`
* `expect_match(x, pattern)` is a shortcut for `expect_true(all(grepl(pattern, n)))` 

etc